compiler/src/iree/compiler/ConstEval/JitGlobals.cpp

```diff
@@ -212,6 +260,58 @@ class ProgramBuilder {
     return m;
   }
 
+  // Returns true if the given |initializerOp| is a constant expression that is
+  // able to be evaluated by this pass.
+  bool isConstExpr(IREE::Util::InitializerOp initializerOp) {
+    // Bail if multiple blocks. It would be possible to support these in theory
+    // but unclear if worth it in practice given the predominance of SCF at the
+    // levels we run things. What we'd require is adding a single exit block
+    // that stored to the globals unconditionally.
+    if (!initializerOp.getBody().hasOneBlock()) {
+      emitWarning(initializerOp.getLoc())
+          << "skipping consteval initializer: initializers with >1 block not "
+             "yet supported";
+      return false;
+    }
+
+    // Don't allow anything with parameters today. We could handle these by
+    // passing in the parameter file for use but would likely also want to bind
+    // a writeable parameter file to produce into.
+    if (isParameterized(sourceSymbolTable, initializerOp)) {
+      emitWarning(initializerOp.getLoc())
+          << "skipping consteval initializer: uses parameters";
+      return false;
+    }
+
+    // Today we require that all values are constant expressions. We could slice
+    // out just the ones that are.
+    for (auto &op : initializerOp.getOps()) {
+      // DO NOT SUBMIT
+      op.dump();
+      // Only allow stores to immutable globals (ones we initialize).
+      if (isa<IREE::Util::ReturnOp>(op)) {
+        continue;
+      } else if (auto storeOp =
+                     dyn_cast<IREE::Util::GlobalStoreOpInterface>(op)) {
+        auto globalOp = sourceSymbolTable.lookup<IREE::Util::GlobalOpInterface>(
+            storeOp.getGlobalAttr().getAttr());
+        if (!globalOp || globalOp.isGlobalMutable()) {
+          emitWarning(initializerOp.getLoc())
+              << "skipping consteval initializer: has mutable global store";
+          return false;
+        }
+        continue;
+      } else if (!constExprAnalysis.isConstExprOperation(&op)) {
```

Comment:
It's not clear to me this is the right check. If there is something like `arith.extf` on a large constant, the analysis won't say that is a "constant expression," even though it is because ConstExprHoisting is trying to figure out what should go in an initializer, not what is JITable.

With that said, it should work for any initializers formed by the hoisting pass, but it does feel like a strange assumption that what we choose to hoist == what we choose to JIT here.

Reply:
yeah, I think it's wrong - I'm not quite sure what's right, though :P
I want the analysis to say it's a constexpr. I'm not using the policy (which is what has maximum sizes/etc) and just looking to rely on constness (side-effect free and derived only from other constants). I think the analysis has a weird quirk around root ops (so tensor.empty isn't a const expr, etc) that I need to figure out.

Reply:
I also suspect that it is not modeling things like encodings properly right now either. A `linalg_ext.set_encoding` is hoistable but is not const-evaluatable. Maybe some kind of side-effecting optrait?